src/main.rs

```diff
@@ -104,67 +108,98 @@ fn main() -> Result<()> {
+            anyhow!(
+                "only one interface descriptor is expected, got:\n{:#?}",
+                intf.descriptors().format("\n")
+            )
         })?;
+        ensure!(
+            intf_desc.class_code() == 0x03
+                && intf_desc.sub_class_code() == 0x00
+                && intf_desc.protocol_code() == 0x00,
+            "unexpected interface parameters: {:#?}",
+            intf_desc
+        );
+
+        // Look for suitable endpoints
+        if let Some(endpt_desc) = intf_desc.endpoint_descriptors().find(|ep| {
+            ep.transfer_type() == TransferType::Interrupt && ep.address() == endpoint_addr
+        }) {
+            debug!("Found endpoint {endpt_desc:?}");
+            return Ok((iface_num, endpt_desc.address()));
+        }
+    }
+
+    Err(anyhow!("No valid interface/endpoint combination found!"))
+}
+
+fn open_keyboard(options: &Options) -> Result<Box<dyn Keyboard>> {
+    // Find USB device based on the product id
+    let (device, desc, id_product) = find_device(options).context("find USB device")?;
+
     ensure!(
-        intf_desc.class_code() == 0x03
-            && intf_desc.sub_class_code() == 0x00
-            && intf_desc.protocol_code() == 0x00,
-        "unexpected interface parameters: {:#?}", intf_desc
+        desc.num_configurations() == 1,
+        "only one device configuration is expected"
     );
 
-    let mut endpt_descs = intf_desc
-        .endpoint_descriptors()
-        .filter(|ep| ep.transfer_type() == TransferType::Interrupt);
-    let endpt_desc = if let Some(endpoint_address) = options.devel_options.endpoint_address {
-        endpt_descs
-            .find(|d| d.address() == endpoint_address)
-            .ok_or_else(|| anyhow!("endpoint with address {} not found", endpoint_address))?
-    } else {
-        endpt_descs
-            .exactly_one()
-            .map_err(|_| {
-                anyhow!(indoc!(
-                    "single interrupt endpoint is expected, got:
-                    {:#?}
-
-                    You may try to choose one using --endpoint-address"
-                ), intf_desc.endpoint_descriptors().format("\n"))
-            })?
-    };
+    // Find correct endpoint
+    let (intf_num, endpt_addr) = find_interface_and_endpoint(
+        &device,
+        options.devel_options.interface_number,
+        options.devel_options.endpoint_address,
+    )?;
 
     // Open device.
     let mut handle = device.open().context("open USB device")?;
     let _ = handle.set_auto_detach_kernel_driver(true);
-    handle.claim_interface(intf.number()).context("claim interface")?;
-    Keyboard::new(handle, endpt_desc.address()).context("init keyboard")
+    handle
+        .claim_interface(intf_num)
+        .context("claim interface")?;
+
+    match id_product {
+        0x8840 | 0x8842 => {
+            k884x::Keyboard884x::new(handle, endpt_addr).map(|v| Box::new(v) as Box<dyn Keyboard>)
+        }
+        0x8880 => {
```

Comment:
Right now fixed product ID is 0x8890, but it is no longer supported, why?

Reply:
A typo which got propagated and eventually became the truth. Fixed.